internal/runner/run_result.go

```diff
@@ -0,0 +1,155 @@
+	for _, testCase := range testCases {
+		identifier := testIdentifier(testCase)
+		r.mutedTestLookup[identifier] = true
+	}
+}
+
+func NewRunResult(mutedTests []plan.TestCase) *RunResult {
+	r := &RunResult{
+		tests:           make(map[string]*TestResult),
+		mutedTestLookup: make(map[string]bool),
+	}
+
+	for _, testCase := range mutedTests {
+		identifier := testIdentifier(testCase)
+		r.mutedTestLookup[identifier] = true
+	}
+	return r
+}
+
+// getTest finds or creates a TestResult struct for a given test case
+// in the tests map, and returns a pointer to it.
+func (r *RunResult) getTest(testCase plan.TestCase) *TestResult {
+	if r.tests == nil {
+		r.tests = make(map[string]*TestResult)
+	}
+
+	testIdentifier := testIdentifier(testCase)
+
+	test, exists := r.tests[testIdentifier]
+	if !exists {
+		test = &TestResult{
+			TestCase: testCase,
+		}
+		r.tests[testIdentifier] = test
+	}
+	return test
+}
+
+// RecordTestResult records the result of a test case.
+// If the test case found in the mutedTestLookup, it will be marked as muted.
+func (r *RunResult) RecordTestResult(testCase plan.TestCase, status TestStatus) {
+	test := r.getTest(testCase)
+	test.Status = status
+	test.ExecutionCount++
+	if r.mutedTestLookup[testIdentifier(testCase)] {
+		test.Muted = true
+	}
+}
+
+// FailedTests returns a list of test cases that failed.
+func (r *RunResult) FailedTests() []plan.TestCase {
+	var failedTests []plan.TestCase
+
+	for _, test := range r.tests {
+		if test.Status == TestStatusFailed && !test.Muted {
+			failedTests = append(failedTests, test.TestCase)
+		}
+	}
+
+	return failedTests
+}
+
+func (r *RunResult) MutedTests() []TestResult {
+	var mutedTests []TestResult
+	for _, test := range r.tests {
+		if test.Muted {
+			mutedTests = append(mutedTests, *test)
+		}
+	}
+
+	return mutedTests
+}
+
+// Status returns the overall status of the test run.
+// If there is an error, it returns RunStatusError.
+// If there are failed tests, it returns RunStatusFailed.
+// Otherwise, it returns RunStatusPassed.
+func (r *RunResult) Status() RunStatus {
+	if r.err != nil {
+		return RunStatusError
+	}
+
+	if len(r.FailedTests()) > 0 {
+		return RunStatusFailed
+	}
+
+	return RunStatusPassed
+}
+
+type RunStatistics struct {
+	Total            int
+	PassedOnFirstRun int
+	PassedAfterRetry int
+	MutedPassed      int
+	MutedFailed      int
+	Failed           int
+}
+
+func (r *RunResult) Statistics() RunStatistics {
+	var passedOnFirstRun, passedAfterRetry, mutedPassed, mutedFailed, failed int
+
+	for _, testResult := range r.tests {
+		if testResult.Muted {
+			if testResult.Status == TestStatusPassed {
+				mutedPassed++
+			} else if testResult.Status == TestStatusFailed {
+				mutedFailed++
+			}
+		} else if testResult.Status == TestStatusPassed {
+			if testResult.ExecutionCount > 1 {
+				passedAfterRetry++
+			} else {
+				passedOnFirstRun++
+			}
+		} else if testResult.Status == TestStatusFailed {
+			failed++
+		}
```

Comment:
I would lean more on `switch` here, even when the conditions are comparing different things; `switch` checks cases sequentially:

```suggestion
		switch {
		case testResult.Muted:
			switch testResult.Status {
			case TestStatusPassed:
				mutedPassed++
			case TestStatusFailed:
				mutedFailed++
			}
			
		case testResult.Status == TestStatusPassed:
			if testResult.ExecutionCount > 1 {
				passedAfterRetry++
			} else {
				passedOnFirstRun++
			}
			
		case testResult.Status == TestStatusFailed:
			failed++
		}
```